packages/native-stack/src/views/NativeStackView.native.tsx

```diff
@@ -525,99 +544,132 @@ export function NativeStackView({ state, navigation, descriptors }: Props) {
+
+            const isModal = modalRouteKeys.includes(route.key);
+
+            return (
+              <SceneView
+                key={route.key}
+                index={index}
+                focused={isFocused}
+                descriptor={descriptor}
+                previousDescriptor={previousDescriptor}
+                nextDescriptor={nextDescriptor}
+                screenRefs={screensRefs}
+                isPresentationModal={isModal}
+                onWillDisappear={() => {
+                  navigation.emit({
+                    type: 'transitionStart',
+                    data: { closing: true },
+                    target: route.key,
+                  });
+                }}
+                onWillAppear={() => {
+                  navigation.emit({
+                    type: 'transitionStart',
+                    data: { closing: false },
+                    target: route.key,
+                  });
+                }}
+                onAppear={() => {
+                  navigation.emit({
+                    type: 'transitionEnd',
+                    data: { closing: false },
+                    target: route.key,
+                  });
+                }}
+                onDisappear={() => {
+                  navigation.emit({
+                    type: 'transitionEnd',
+                    data: { closing: true },
+                    target: route.key,
+                  });
+                }}
+                onDismissed={(event) => {
+                  navigation.dispatch({
+                    ...StackActions.pop(event.nativeEvent.dismissCount),
+                    source: route.key,
+                    target: state.key,
+                  });
+
+                  setNextDismissedKey(route.key);
+                }}
+                onHeaderBackButtonClicked={() => {
+                  navigation.dispatch({
+                    ...StackActions.pop(),
+                    source: route.key,
+                    target: state.key,
+                  });
+                }}
+                onNativeDismissCancelled={(event) => {
+                  navigation.dispatch({
+                    ...StackActions.pop(event.nativeEvent.dismissCount),
+                    source: route.key,
+                    target: state.key,
+                  });
+                }}
+                onGestureCancel={() => {
+                  navigation.emit({
+                    type: 'gestureCancel',
+                    target: route.key,
+                  });
+                }}
+              />
+            );
+          })}
+        </ScreenStack>
+      </ScreenGestureDetector>
     </SafeAreaProviderCompat>
   );
 }
 
+const ScreenWithRef = React.forwardRef<InnerScreen & View, ScreenProps>(
+  (props, ref) => {
+    const ScreenWrapper = React.useContext(ScreenContext) || InnerScreen;
+
+    return <ScreenWrapper {...props} ref={ref} />;
+  }
+);
```

Comment:
We shouldn't be duplicating this logic into native stack code. Adding `React.forwardRef` here should be enough: https://github.com/software-mansion/react-native-screens/blob/c2ac40f5b3b1cc1c152126a1dd1c42b6834b8f74/src/components/Screen.tsx#L207

Exposing `InnerScreen` and requiring consumer code to use the context because of the logic inside screens leaks internal implementation details - which makes harder for you to change how things work in the future. It also makes the code for consumers much more verbose (e.g. a lot of code here instead of one line change in `react-native-screens`).